### PR TITLE
Add implementation of DSL rewrite database

### PR DESCRIPTION
**Pull request by ajreynol**
This primarily sets up indexing and caching of rewrite rules, which will be used by the proof reconstruction algorithm.

### CODE REVIEW

File: src/rewriter/rewrite_db.cpp [hunk 1]
```diff
@@ -36,8 +53,140 @@ void RewriteDb::addRule(DslPfRule id,
                         Node context,
                         bool isFlatForm)
 {
+  NodeManager* nm = NodeManager::currentNM();
+  // flatten if necessary, and if possible
```
Comment (HanielB) [inline src/rewriter/rewrite_db.cpp:57]: this seems out of place

File: src/rewriter/rewrite_db.h [hunk 2]
```diff
@@ -60,6 +73,46 @@ class RewriteDb
                Node cond,
                Node context,
                bool isFlatForm = false);
```
Comment (HanielB) [inline src/rewriter/rewrite_db.h:75]: can you clarify what being in flat form means?

**Review by HanielB — APPROVED** [review]
LGTM, just minor comments

File: src/rewriter/rewrite_db.cpp (hunk 1 above)
Comment (ajreynol) [reply src/rewriter/rewrite_db.cpp:57]: This is a misplaced comment for a feature that was removed, I'll remove.

File: src/rewriter/rewrite_db.h (hunk 2 above)
Comment (ajreynol) [reply src/rewriter/rewrite_db.h:75]: It's a deprecated feature, I'll remove.

**Pull request merged**